### PR TITLE
support aap through vip crd using selector

### LINKED ISSUE #3187: subnet spec vips vs  vip crd
image

按照当前的代码来看，这两者的 vip 基本功能是一样的

- vip crd 本身尚未实现 安全组下的 AAP （https://docs.openstack.org/developer/dragonflow/specs/allowed_address_pairs.html）
- subnet 的 vip 是将 vip 作为所有 lsp 的 AAP

后续可以 vip crd 支持 app 后，或许 可以考虑移除这个逻辑

### LINKED ISSUE #3318: vip will use selector to locate where its parent lsp
### Feature request

image

pod use selector to locate its node

VIP use selector to locate its parents lsp

### PR DESCRIPTION
**Pull request by zcq98**
- [x] Make sure you have followed [Kube-OVN Code Style](https://github.com/kubeovn/kube-ovn/blob/master/CODE_STYLE.md).

### What type of this PR
Examples of user facing changes:
- Features

1、vip supports both manual allocation and automatic allocation.
2、adding or deleting vips through subnet.spec.vips will update the subnet, configuring aap through vip crd avoids this problem and is easy to manage vips.
3、vip crd supports selector. Users can flexibly configure whether to enable aap through label. The same pod supports multiple aaps.

### Which issue(s) this PR fixes:
Fixes #3318 
Fixes #3187 

### <samp>🤖 Generated by Copilot at c0f259f</samp>

This pull request adds a feature to support virtual IP for vlan in kube-ovn. It introduces the `selector` field in the CRD schemas of `Vlan` and `Vip` resources, and updates the controller logic to handle pod label changes and pod deletion events for virtual IP. It also modifies the `VipSpec` struct in `pkg/apis/kubeovn/v1/types.go` to include the `selector` field.

### <samp>🤖 Generated by Copilot at c0f259f</samp>

> _Oh we are the coders of the `kubeovn` crew_
> _And we've got a feature that is shiny and new_
> _We'll add a `selector` field to the `Vip` and `Vlan`_
> _And we'll heave away and haul on the count of three_

### <samp>🤖 Generated by Copilot at c0f259f</samp>

*  Add `selector` field to `VlanSpec` and `VipSpec` schemas to specify pod label selectors for vlan and virtual IP ([link](https://github.com/kubeovn/kube-ovn/pull/3359/files?diff=unified&w=0#diff-d671263a0aa28dac78faa045e43d18cde05bf81846c6d1bf1db009cddda883f2R1493-R1496), [link](https://github.com/kubeovn/kube-ovn/pull/3359/files?diff=unified&w=0#diff-d671263a0aa28dac78faa045e43d18cde05bf81846c6d1bf1db009cddda883f2R1539-R1542), [link](https://github.com/kubeovn/kube-ovn/pull/3359/files?diff=unified&w=0#diff-7b60edb5fc7c18c120f5a0077606604cfb6be9bb55517095bb9380877ef716c5R1804-R1807), [link](https://github.com/kubeovn/kube-ovn/pull/3359/files?diff=unified&w=0#diff-7b60edb5fc7c18c120f5a0077606604cfb6be9bb55517095bb9380877ef716c5R1850-R1853))
*  Add `Selector` field to `VipSpec` struct in `pkg/apis/kubeovn/v1/types.go` to represent pod label selectors for virtual IP ([link](https://github.com/kubeovn/kube-ovn/pull/3359/files?diff=unified&w=0#diff-69f75d53361877cc02ab6f413f306486d29287c6400b92b71c4c74f018f3234aR811))
*  Add `reflect` and `slices` packages to `pkg/controller/vip.go` and `pkg/controller/pod.go` to provide functions for comparing values and working with slices ([link](https://github.com/kubeovn/kube-ovn/pull/3359/files?diff=unified&w=0#diff-449e8d1bd46b9c978e208dde912a1d7a76d17ce553067f8d0a67a27f17338d26R9), [link](https://github.com/kubeovn/kube-ovn/pull/3359/files?diff=unified&w=0#diff-c53f89147c39d7d4c3d3046d288a0cea09719776e947bfb3ef01ea29f75b1fb0R8-R9))
*  Add `syncVirtualParents` function to `pkg/controller/vip.go` to create and update virtual port and its virtual parents based on virtual IP selector ([link](https://github.com/kubeovn/kube-ovn/pull/3359/files?diff=unified&w=0#diff-c53f89147c39d7d4c3d3046d288a0cea09719776e947bfb3ef01ea29f75b1fb0R386-R446))
*  Call `syncVirtualParents` function in `handleAddVirtualIP` and `handleUpdateVirtualIP` functions in `pkg/controller/vip.go` if virtual IP has a v4ip assigned ([link](https://github.com/kubeovn/kube-ovn/pull/3359/files?diff=unified&w=0#diff-c53f89147c39d7d4c3d3046d288a0cea09719776e947bfb3ef01ea29f75b1fb0R251-R255), [link](https://github.com/kubeovn/kube-ovn/pull/3359/files?diff=unified&w=0#diff-c53f89147c39d7d4c3d3046d288a0cea09719776e947bfb3ef01ea29f75b1fb0R268-R272))
*  Delete virtual port associated with deleted virtual IP from logical switch in `handleDelVirtualIP` function in `pkg/controller/vip.go` ([link](https://github.com/kubeovn/kube-ovn/pull/3359/files?diff=unified&w=0#diff-c53f89147c39d7d4c3d3046d288a0cea09719776e947bfb3ef01ea29f75b1fb0R327-R332))
*  Enqueue virtual IP for update if `selector` field has changed in `enqueueUpdateVirtualIP` function in `pkg/controller/vip.go` ([link](https://github.com/kubeovn/kube-ovn/pull/3359/files?diff=unified&w=0#diff-c53f89147c39d7d4c3d3046d288a0cea09719776e947bfb3ef01ea29f75b1fb0L47-R50))
*  List and enqueue virtual IPs that belong to the same subnet as the updated or deleted pod and match the pod labels or are virtual parents of the virtual port in `handleUpdatePod` and `handleDeletePod` functions in `pkg/controller/pod.go` ([link](https://github.com/kubeovn/kube-ovn/pull/3359/files?diff=unified&w=0#diff-449e8d1bd46b9c978e208dde912a1d7a76d17ce553067f8d0a67a27f17338d26R257-R292), [link](https://github.com/kubeovn/kube-ovn/pull/3359/files?diff=unified&w=0#diff-449e8d1bd46b9c978e208dde912a1d7a76d17ce553067f8d0a67a27f17338d26L1050-R1107))

### CODE REVIEW

File: pkg/controller/vip.go [hunk 1]
```diff
@@ -311,6 +324,12 @@ func (c *Controller) handleDelVirtualIP(vip *kubeovnv1.Vip) error {
 			return err
 		}
 	}
+	// delete virtual ports
+	lspName := fmt.Sprintf("%s-vip-%s", vip.Spec.Subnet, vip.Status.V4ip)
```
Comment (bobz965) [inline pkg/controller/vip.go:328]: 直接用crd的名字会更好些

File: pkg/controller/pod.go [hunk 2]
```diff
@@ -253,6 +254,42 @@ func (c *Controller) enqueueDeletePod(obj interface{}) {
...
+
+	if !reflect.DeepEqual(oldPod.Labels, newPod.Labels) {
+		vips, err := c.virtualIpsLister.List(labels.SelectorFromSet(labels.Set{
+			util.SubnetNameLabel: newPod.Annotations[util.LogicalSwitchAnnotation],
+		}))
+		if err != nil {
+			klog.Error(err)
+			return
+		}
+		for _, vip := range vips {
```
Comment (bobz965) [inline pkg/controller/pod.go:266]: vip 和 pod 可以基于label再建立一些更精确的对应关系，缩小vip的for loop开销。
vip 可以直接 enqueue
vip 关于 virtualParents 的对比逻辑这部分逻辑是省不掉的，放在 pod 这里会有一些代码重复。
如果担心 vip 的 virtualParents 用 updateVirtualIPQueue 有一些无关逻辑重复执行，可以单独搞一个 updateVirtualIPParentsQueue 单独针对性的处理这种逻辑。

File: pkg/controller/vip.go [hunk 3]
```diff
@@ -364,6 +383,67 @@ func (c *Controller) acquireIPAddress(subnetName, name, nicName string) (string,
...
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+
+	// add new virtual port
+	if err = c.OVNNbClient.CreateVirtualLogicalSwitchPorts(cachedVip.Spec.Subnet, cachedVip.Status.V4ip); err != nil {
+		klog.Errorf("create virtual port with vips %v from logical switch %s: %v", cachedVip.Status.V4ip, cachedVip.Spec.Subnet, err)
```
Comment (bobz965) [inline pkg/controller/vip.go:397]: 字符串用%s

File: pkg/controller/vip.go [hunk 4]
```diff
@@ -364,6 +383,67 @@ func (c *Controller) acquireIPAddress(subnetName, name, nicName string) (string,
...
+		return err
+	}
+
+	selectors := make([]map[string]string, 0)
+	for _, v := range cachedVip.Spec.Selector {
```
Comment (bobz965) [inline pkg/controller/vip.go:402]: 感觉这块可以移除

File: pkg/controller/vip.go [hunk 5]
```diff
@@ -364,6 +383,67 @@ func (c *Controller) acquireIPAddress(subnetName, name, nicName string) (string,
...
+		}
+
+		for _, pod := range pods {
+			if pod.Annotations[util.LogicalSwitchAnnotation] != cachedVip.Spec.Subnet {
+				continue
+			}
+			key := fmt.Sprintf("%s/%s", pod.Namespace, pod.Name)
+			ports, err := c.OVNNbClient.ListNormalLogicalSwitchPorts(true, map[string]string{"pod": key})
+			if err != nil {
+				klog.Errorf("failed to list lsps of pod '%s', %v", pod.Name, err)
```
Comment (bobz965) [inline pkg/controller/vip.go:428]:
```suggestion
				klog.Errorf("failed to list lsps of pod '%s', %v", key, err)
```

File: pkg/controller/pod.go [hunk 6]
```diff
@@ -253,6 +254,42 @@ func (c *Controller) enqueueDeletePod(obj interface{}) {
 func (c *Controller) enqueueUpdatePod(oldObj, newObj interface{}) {
 	oldPod := oldObj.(*v1.Pod)
 	newPod := newObj.(*v1.Pod)
+
+	if !reflect.DeepEqual(oldPod.Labels, newPod.Labels) {
+		vips, err := c.virtualIpsLister.List(labels.SelectorFromSet(labels.Set{
+			util.SubnetNameLabel: newPod.Annotations[util.LogicalSwitchAnnotation],
```
Comment (lynn901) [inline pkg/controller/pod.go:260]: 仅对ovn网络做了处理，未考虑multus网络

File: pkg/controller/pod.go [hunk 7]
```diff
@@ -255,36 +255,23 @@ func (c *Controller) enqueueUpdatePod(oldObj, newObj interface{}) {
...
-				klog.Error(err)
-				return
-			}
-			if lsp != nil {
-				virtualParents := strings.Split(lsp.Options["virtual-parents"], ",")
-				if slices.Contains(virtualParents, fmt.Sprintf("%s.%s", newPod.Name, newPod.Namespace)) {
-					c.updateVirtualIPQueue.Add(vip.Name)
-					continue
+	if !reflect.DeepEqual(oldPod.Labels, newPod.Labels) || !reflect.DeepEqual(oldPod.Annotations, newPod.Annotations) {
+		oldAaps := strings.Split(oldPod.Annotations[util.AAPAnnotation], ",")
```
Comment (bobz965) [inline pkg/controller/pod.go:259]:
```suggestion
		oldAAPs := strings.Split(oldPod.Annotations[util.AAPAnnotation], ",")
```

File: pkg/controller/vip.go [hunk 8]
```diff
@@ -391,53 +388,51 @@ func (c *Controller) syncVirtualParents(key string) error {
...
+	// add new virtual port if not exist
 	if err = c.OVNNbClient.CreateVirtualLogicalSwitchPorts(cachedVip.Spec.Subnet, cachedVip.Status.V4ip); err != nil {
-		klog.Errorf("create virtual port with vips %v from logical switch %s: %v", cachedVip.Status.V4ip, cachedVip.Spec.Subnet, err)
+		klog.Errorf("create virtual port with vips %s from logical switch %s: %v", cachedVip.Status.V4ip, cachedVip.Spec.Subnet, err)
 		return err
 	}
 
-	selectors := make([]map[string]string, 0)
+	selectors := make(map[string]string)
 	for _, v := range cachedVip.Spec.Selector {
```
Comment (bobz965) [inline pkg/controller/vip.go:402]:
``` 
sel, err := metav1.ParseToLabelSelector(strings.Join(cachedVip.Spec.Selector, ","))
if err != nil {
    klog.Errorf("failed to parse label selector, %v", err)
    return err
}

```

File: pkg/controller/vip.go [hunk 9]
```diff
@@ -364,6 +413,65 @@ func (c *Controller) acquireIPAddress(subnetName, name, nicName string) (string,
...
+			if podNet.Subnet.Name == cachedVip.Spec.Subnet {
+				portName := ovs.PodNameToPortName(pod.Name, pod.Namespace, podNet.Subnet.Spec.Provider)
+				virtualParents = append(virtualParents, portName)
+				break
+			}
+		}
+	}
+
+	if err = c.OVNNbClient.SetLogicalSwitchPortVirtualParents(cachedVip.Spec.Subnet, strings.Join(virtualParents, ","), cachedVip.Status.V4ip); err != nil {
+		klog.Errorf("set vip %s virtual parents %s: %v", cachedVip.Status.V4ip, virtualParents, err)
```
Comment (bobz965) [inline pkg/controller/vip.go:468]:
```suggestion
		klog.Errorf("set vip %s virtual parents %s: %v", cachedVip.Name, virtualParents, err)
```

**Comment by bobz965** [comment]
please attach the docs

File: pkg/controller/vip.go [hunk 10]
```diff
@@ -426,8 +424,8 @@ func (c *Controller) handleUpdateVirtualParents(key string) error {
 		return nil
 	}
 	// add new virtual port if not exist
-	if err = c.OVNNbClient.CreateVirtualLogicalSwitchPorts(cachedVip.Spec.Subnet, cachedVip.Status.V4ip); err != nil {
-		klog.Errorf("create virtual port with vips %s from logical switch %s: %v", cachedVip.Status.V4ip, cachedVip.Spec.Subnet, err)
+	if err = c.OVNNbClient.CreateVirtualLogicalSwitchPort(cachedVip.Name, cachedVip.Spec.Subnet, cachedVip.Status.V4ip); err != nil {
+		klog.Errorf("create virtual port with vip %s from logical switch %s: %v", cachedVip.Status.V4ip, cachedVip.Spec.Subnet, err)
```
Comment (bobz965) [inline pkg/controller/vip.go:428]: vip %s --> use name

File: pkg/controller/vip.go [hunk 11]
```diff
@@ -464,8 +462,8 @@ func (c *Controller) handleUpdateVirtualParents(key string) error {
 		}
 	}
 
-	if err = c.OVNNbClient.SetLogicalSwitchPortVirtualParents(cachedVip.Spec.Subnet, strings.Join(virtualParents, ","), cachedVip.Status.V4ip); err != nil {
-		klog.Errorf("set vip %s virtual parents %s: %v", cachedVip.Status.V4ip, virtualParents, err)
+	if err = c.OVNNbClient.SetLogicalSwitchPortVirtualParent(cachedVip.Name, strings.Join(virtualParents, ",")); err != nil {
```
Comment (bobz965) [inline pkg/controller/vip.go:465]: parents := strings.Join(virtualParents, ",")
if err = c.OVNNbClient.SetLogicalSwitchPortVirtualParent(cachedVip.Name, parents); err != nil {
klog.Errorf("set vip %s virtual parents %s: %v", cachedVip.Name, parents, err)

File: pkg/ovs/ovn-nb-logical_switch_port.go [hunk 12]
```diff
@@ -185,6 +185,41 @@ func (c *OVNNbClient) CreateVirtualLogicalSwitchPorts(lsName string, ips ...stri
...
+	}
+
+	op, err := c.CreateLogicalSwitchPortOp(lsp, lsName)
+	if err != nil {
+		klog.Error(err)
+		return err
+	}
+
+	if err := c.Transact("lsp-add", op); err != nil {
+		return fmt.Errorf("create virtual logical switch ports for logical switch %s: %v", lsName, err)
```
Comment (bobz965) [inline pkg/ovs/ovn-nb-logical_switch_port.go:217]: return fmt.Errorf("create virtual logical switch port %s for logical switch %s: %v", lspName, lsName, err)

File: pkg/ovs/ovn-nb-logical_switch_port.go [hunk 13]
```diff
@@ -255,6 +290,31 @@ func (c *OVNNbClient) SetLogicalSwitchPortVirtualParents(lsName, parents string,
 	return nil
 }
 
+// CreateVirtualLogicalSwitchPort update one virtual type logical switch port virtual-parents
+func (c *OVNNbClient) SetLogicalSwitchPortVirtualParent(lspName, parents string) error {
```
Comment (bobz965) [inline pkg/ovs/ovn-nb-logical_switch_port.go:294]: SetLogicalSwitchPortVirtualParents

File: pkg/ovs/ovn-nb-logical_switch_port.go [hunk 14]
```diff
@@ -255,6 +290,31 @@ func (c *OVNNbClient) SetLogicalSwitchPortVirtualParents(lsName, parents string,
...
+	lsp, err := c.GetLogicalSwitchPort(lspName, true)
+	if err != nil {
+		klog.Error(err)
+		return fmt.Errorf("get logical switch port %s: %v", lspName, err)
+	}
+
+	lsp.Options["virtual-parents"] = parents
```
Comment (bobz965) [inline pkg/ovs/ovn-nb-logical_switch_port.go:301]: compare before set

**Comment by zcq98** [comment]
docs: https://github.com/kubeovn/docs/pull/135

**Comment by zcq98** [comment]
> please attach the docs

docs: https://github.com/kubeovn/docs/pull/135

File: pkg/util/const.go [hunk 15]
```diff
@@ -21,6 +21,7 @@ const (
 	FipEnableAnnotation  = "ovn.kubernetes.io/enable_fip"
 	FipFinalizer         = "ovn.kubernetes.io/fip"
 	VipAnnotation        = "ovn.kubernetes.io/vip"
+	AAPAnnotation        = "ovn.kubernetes.io/aaps"
```
Comment (bobz965) [inline pkg/util/const.go:24]: AAPsAnnotation

**Comment by bobz965** [comment]
hi @oilbeater , could you help review this ?

**Review by bobz965 — APPROVED** [review]

**Review by bobz965 — APPROVED** [review]

File: pkg/controller/vip.go [hunk 16]
```diff
@@ -364,6 +411,66 @@ func (c *Controller) acquireIPAddress(subnetName, name, nicName string) (string,
...
+	sel, _ := metav1.LabelSelectorAsSelector(&metav1.LabelSelector{MatchLabels: selectors})
+	pods, err := c.podsLister.Pods(cachedVip.Spec.Namespace).List(sel)
+	if err != nil {
+		klog.Errorf("failed to list pods that meet selector requirements, %v", err)
+		return err
+	}
+
+	var virtualParents []string
+	for _, pod := range pods {
+		if aaps := strings.Split(pod.Annotations[util.AAPsAnnotation], ","); !slices.Contains(aaps, cachedVip.Name) {
```
Comment (oilbeater) [inline pkg/controller/vip.go:449]: should check if pod.Annotations is nil to prevent panic

File: pkg/controller/vip.go [hunk 17]
```diff
@@ -364,6 +411,66 @@ func (c *Controller) acquireIPAddress(subnetName, name, nicName string) (string,
...
+			return nil
+		}
+		return err
+	}
+	// only pods in the same namespace as vip are allowed to use aap
+	if cachedVip.Status.V4ip == "" || cachedVip.Spec.Namespace == "" {
+		return nil
+	}
+	// add new virtual port if not exist
+	if err = c.OVNNbClient.CreateVirtualLogicalSwitchPort(cachedVip.Name, cachedVip.Spec.Subnet, cachedVip.Status.V4ip); err != nil {
```
Comment (oilbeater) [inline pkg/controller/vip.go:427]: When this port will be deleted

**Review by oilbeater — CHANGES_REQUESTED** [review]

File: pkg/controller/vip.go (hunk 17 above)
Comment (zcq98) [reply pkg/controller/vip.go:427]: > When this port will be deleted

this port will be deleted when vip is deleted

**Review by oilbeater — APPROVED** [review]

**Pull request merged**